### ISSUE TITLE
flutter run doesn't handle failure on the device during file upload

### ISSUE THREAD

**Issue by Hixie** [flutter/flutter#6495]
```
Exception from flutter run: HttpException: Connection closed before full header was received, uri = http://127.0.0.1:8100
dart:io                                          _HttpClient.putUrl
package:flutter_tools/src/devfs.dart 225         _DevFSHttpWriter._scheduleWrite.<async>
===== asynchronous gap ===========================
dart:async                                       Future.Future.microtask
package:flutter_tools/src/devfs.dart             _DevFSHttpWriter._scheduleWrite
package:flutter_tools/src/devfs.dart 218         _DevFSHttpWriter._scheduleWrites
package:flutter_tools/src/devfs.dart 206         _DevFSHttpWriter.write.<async>
===== asynchronous gap ===========================
dart:async                                       Future.Future.microtask
package:flutter_tools/src/devfs.dart             _DevFSHttpWriter.write
package:flutter_tools/src/devfs.dart 387         DevFS.update.<async>
dart:async                                       _SyncCompleter.complete
package:flutter_tools/src/devfs.dart 540         DevFS._scanDirectory.<async>
===== asynchronous gap ===========================
dart:async                                       _asyncThenWrapperHelper
package:flutter_tools/src/devfs.dart             DevFS.update
package:flutter_tools/src/hot.dart 357           HotRunner._updateDevFS.<async>
===== asynchronous gap ===========================
dart:async                                       Future.Future.microtask
package:flutter_tools/src/hot.dart               HotRunner._updateDevFS
package:flutter_tools/src/hot.dart 290           HotRunner._run.<async>
dart:async                                       _SyncCompleter.complete
package:flutter_tools/src/hot.dart 412           HotRunner._launchFromDevFS.<async>
dart:async                                       _SyncCompleter.complete
package:flutter_tools/src/vmservice.dart 937     FlutterView.runFromSource.<async>
===== asynchronous gap ===========================
dart:async                                       _asyncThenWrapperHelper
package:flutter_tools/src/hot.dart               HotRunner._run
package:flutter_tools/src/hot.dart 150           HotRunner.run.<fn>
package:stack_trace                              Chain.capture
package:flutter_tools/src/hot.dart 149           HotRunner.run
package:flutter_tools/src/commands/run.dart 221  RunCommand.runCommand.<async>
```

**Comment by Hixie** [comment]
cc @johnmccutchan

**Comment by johnmccutchan** [comment]
This was fixed before I left for Munich. But reverted by @tvolkert due to an unrelated iOS breakage. @tvolkert can I reland my fixes?

**Comment by tvolkert** [comment]
The fix has already been re-applied in #6472

**Comment by johnmccutchan** [comment]
Excellent. Thanks!
